src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,399 @@
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Test for two-byte preambles
+            if (bytesRead > 1)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1]);
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Check for binary
+            string initialBytesAsAscii = System.Text.Encoding.ASCII.GetString(initialBytes, 0, bytesRead);
+            if (initialBytesAsAscii.IndexOfAny(nonPrintableCharacters) >= 0)
+            {
+                return FileEncoding.Byte;
+            }
+
+            // we couldn't determine anything from direct examination,
+            // return UTF8 without a BOM which should be good for both Windows and Non-Windows
+            return FileEncoding.UTF8NoBOM;
+        }
+
+        /// <summary>
+        /// Retrieve the PSDefaultFileEncoding preference value if set
+        /// <summary>
+        public static FileEncoding GetEncodingPreference(SessionState sessionState)
+        {
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
+                object tmp = sessionState.PSVariable.GetValue("PSDefaultFileEncoding");
+                LanguagePrimitives.TryConvertTo<FileEncoding>(tmp, out encodingPreference);
+            }
+            catch
+            {
+                ;
+            }
+            return encodingPreference;
+        }
+
+        /// <summary>
+        /// Retrieve the encoding in a provider context
+        /// </summary>
+        public static Encoding GetProviderEncoding(CmdletProvider provider, FileEncoding encoding)
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = GetEncodingPreference(provider.SessionState);
+            // If the encoding isn't set, but is available as $PSDefaultFileEncoding, use that
+            // It the encoding is set use that, otherwise return the default encoding
+            if ( encoding == FileEncoding.Unknown && encodingPreference != FileEncoding.Unknown )
+            {
+                resolvedEncoding = GetEncoding(encodingPreference);
+            }
+            else if ( encoding != FileEncoding.Unknown )
+            {
+                resolvedEncoding = GetEncoding(encoding);
+            }
+            return resolvedEncoding;
+        }
+
+        // This is the way the encoding is implemented in PowerShell 5 and earlier.
+        // If the user sets the default encoding to WindowsLegacy, we will
+        // be able to encode for that
+        internal static Dictionary<String, Encoding> legacyEncodingMap =
+            new Dictionary<string, Encoding>(StringComparer.OrdinalIgnoreCase)
```

Comment:
We don't want to introduce a new circular dependency, but we have implicit knowledge encoded here, either via a string or `typeof` - there is little difference.

Actually removing the implicit knowledge probably isn't worth it, but you could do that via registration - each of these cmdlets would call an internal api to register itself.